packages/ui/src/contexts/MultiProxyContext.tsx

```diff
@@ -64,136 +76,116 @@ const MultiProxyContextProvider = ({ children }: MultisigContextProps) => {
-              .get(delegator.address)
-              ?.multisigs.map(({ address }) => address)
-
-            // finding all the accounts that are pure proxy and that don't include this multisig already
-            if (delegator?.isPureProxy && !currentMultisigsForProxy?.includes(account.address)) {
-              pureProxyAddresses.push({
-                pureAddress: delegator.address,
-                type: type
-              })
-            }
-          })
+        // find the delegatee that are multisigs and put all the infos right away
+        account.delegatorFor.forEach(({ delegatee, type }) => {
+          if (delegatee.isMultisig) {
+            const previousMultisigsForProxy = pureProxyMap.get(account.address)?.multisigs || []
 
-        // if this account is a multisig and is the delegatee for at least a pureProxy
-        if (account.isMultisig && pureProxyAddresses?.length > 0) {
-          pureProxyAddresses.forEach(({ pureAddress, type }) => {
-            const previousMultisigsForProxy = pureProxyMap.get(pureAddress)?.multisigs || []
+            const isAlreadyInMultisigList = !!previousMultisigsForProxy.find(
+              ({ address }) => address === delegatee.address
+            )
+
+            // do not add a second time a multisig
+            if (isAlreadyInMultisigList) return
 
             const newMultisigForProxy = {
-              address: account.address,
-              signatories: getSignatoriesFromAccount(account.signatories),
-              threshold: account?.threshold || undefined,
+              address: delegatee.address,
+              signatories: getSignatoriesFromAccount(delegatee.signatories),
+              threshold: delegatee?.threshold || undefined,
               type
             }
 
-            // add this pureProxy to the Map
-            pureProxyMap.set(pureAddress, {
+            pureProxyMap.set(account.address, {
               multisigs: [...previousMultisigsForProxy, newMultisigForProxy]
             })
-          })
-        } else if (account.isMultisig && pureProxyAddresses.length === 0) {
-          // if this multisig doesn't have a proxy
-          res.push({
-            proxy: undefined,
-            multisigs: [
-              {
-                address: account.address,
-                signatories: getSignatoriesFromAccount(account.signatories),
-                threshold: account.threshold
-              }
-            ]
-          } as MultiProxy)
-        } else {
-          console.error('Unexpected account, it should be a multisig', account)
-        }
+          }
+        })
       })
+    }
 
-      // flatten out proxyMap
-      const proxyArray = Array.from(pureProxyMap.entries()).map(
-        ([proxy, agg]) =>
-          ({
-            proxy,
-            multisigs: agg.multisigs
-          } as MultiProxy)
-      )
-
-      res.push(...proxyArray)
+    // flatten out pureProxyMap
+    const pureProxyArray = Array.from(pureProxyMap.entries()).map(
```

Comment:
again .values() will save on memory allocation

Reply:
absolutely

Reply:
ah not for that one, we need both key and value here.

Reply:
ah true, but why do you need Array.from then?

Reply:
the `.entries()` doesn't return an array, it returns an `IterableIterator<[string, Omit<MultiProxy, "proxy">]>` so you have to make this an array to be able to call `.map` on it.